src/software/util/make_enum_macro/make_enum_macro.h

```diff
@@ -0,0 +1,129 @@
+    }
+}  // namespace
+
+/**
+ * Creates an enum.
+ *
+ * This enum will have support for
+ * 1. Getting the number of enum values
+ * 2. Getting the string representation of each enum value
+ * 3. Getting a list of all enum values
+ *
+ * IMPORTANT: This macro does NOT support enums that have manually specified values
+ * Eg.
+ * MAKE_ENUM(MyEnum, foo=1, bar)
+ * is not valid
+ *
+ * @param name The name of the enum
+ * @param ... The values of the enum
+ */
+#define MAKE_ENUM(name, ...)                                                               \
+    constexpr std::string_view enum_string_args_##name = #__VA_ARGS__;                     \
+    static_assert(                                                                         \
+        isEnumArgsValid(enum_string_args_##name),                                          \
+        "Error: Enums created with the MAKE_ENUM macro may not manually specify enum "     \
+        "values. Please remove any enum assignments with the '=' operator");               \
+    enum class name                                                                      \
+    {                                                                                    \
+        __VA_ARGS__                                                                      \
+        /* Although common convention, we don't use a COUNT value as the last item in */ \
+        /* our enum in order to figure out how many values the enum contains. The */     \
+        /* reason for this is that the compiler includes helpful warnings like */        \
+        /* "enum value not included in switch" to help avoid missing a case in a */      \
+        /* switch statement. Adding an extra COUNT value that will not be handled in */  \
+        /* these cases means we will get warnings (or errors). See this stackoverflow */ \
+        /* answer for a similar explanation: https://stackoverflow.com/a/2102673 */      \
+    }; \
+    inline std::ostream& operator<<(std::ostream& os, name value)                          \
+    {                                                                                      \
+        std::string str                       = #__VA_ARGS__;                              \
+        std::vector<std::string> enum_strings = separateEnumStrings(str);                  \
+        /* This index lookup relies on the assumption that the enum does not manually */   \
+        /* specify any values. If it did, the underlying integer of the given value */     \
+        /* may be out of range of the vector of strings */                                 \
+        os << enum_strings.at(static_cast<int>(value));                                    \
+        return os;                                                                         \
+    }                                                                                      \
+    static inline size_t size##name()                                                      \
```

Comment:
These functions were originally not `static` or `inline`, but that resulted in linker errors for `multiple definitions`.

Adding `inline` fixed this but I didn't think that was a robust solution because `inline` is just a suggestion so if the functions didn't get inlined the linker error would persist.

Therefore I added static which I'm more confident will fix the problem, but would like some second opinions here. Don't want to cause any weird issues down the line.

@garethellis0 @akhilveeraghanta @jonathanlew @amrm3lm 

[Meaning of `static` function in header](https://stackoverflow.com/questions/780730/c-c-static-function-in-header-file-what-does-it-mean)
[Function definitions in header](https://stackoverflow.com/a/14517957) - Seems like function definitions within classes are a special case, and that globally-scoped functions don't get that same implicit inline (maybe).

Reply:
that seems reasonable to me

Reply:
Yeah this seems fine